keps/sig-auth/20200507-external-tls-certificate-authenticator/README.md

```diff
@@ -0,0 +1,749 @@
+-->
+
+This enhancement proposes adding support for authentication via external TLS
+certificate signers, what would enable usage of Hardware Security Modules (HSMs)
+- also known as smartcards, cryptographic processors or, by a popular brand
+name, YubiKeys(tm) via the PKCS#11 standard. This enhancement allows developers
+or automation pipelines to authenticate with the Kubernetes cluster, without
+requiring access to the client key, hence improving compliance and security.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users.  The
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+<!-- How would you react if your laptop was stolen? Are you worried about 
+attackers performing a cold boot attack to extract your Kubernetes 
+credentials?  -->
+Are you worried about someone getting access to your Kubernetes credentials?
+(For example, by extracting private keys from your laptop using malware or by
+performing a cold boot attack.) Do you already use a YubiKey for SSH and GPG,
+and wonder why you cannot use it with kubectl? If yes, then this enhancement is
+for you!
+
+Highly regulated environments, such as FinTech, require delegating all digital
+key operations to specialized [Hardware Security Modules
+(HSMs)](https://en.wikipedia.org/wiki/Hardware_security_module). Amongst others,
+HSMs increase security by storing digital keys without allow them to be
+extracted. Authentication, encryption and signing is performed via a standard
+such as the PKCS#11 on the HSM directly. In fact, many regulated environments
+already require developers and operators to store SSH and GPG keys on the
+[YubiKey](https://en.wikipedia.org/wiki/YubiKey), a popular HSM connected via
+USB.
+
+Unfortunately, as of today, kubectl lacks support for PKCS#11 (see [Issue
+#64783](https://github.com/kubernetes/kubernetes/issues/64783)). Indeed, kubectl
+requires direct access to the client key data, which can either be stored in the
+kubeconfig or provided via a [credentials
+plugin](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#client-go-credential-plugins).
+Because of this, some security bloggers have even argued for [not using
+certificates](https://www.tremolosecurity.com/kubernetes-dont-use-certificates-for-authentication/)
+in kubectl at all.
+
+### Goals
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+- kubectl can authenticate to a Kubernetes cluster with an external TLS
+  certificate signer, for example a PKCS#11-compatible HSM, such as
+  [SoftHSM](https://github.com/opendnssec/SoftHSMv2) or
+  [YubiKey](https://www.yubico.com/)
```

Comment:
This could be mitigated using short-lived (hours or days) certs.
For example, an employee can be required to re-login into their usual corp systems daily, during which a fresh cert is provisioned.

Reply:
Sure, but I am skeptical about this being a common approach.  I have no delved deeply into the YubiKey API, but unless the process can be fully automated, certs are likely to be provisioned with a long life span.